### PR TITLE
New Resource: `azurerm_policy_definition`

### PR DESCRIPTION
**Pull request by jaymitre**
The goal of this change is to allow CRUD operations to be performed on azure policies using terraform.

### CODE REVIEW

**Comment by jaymitre** [comment]
Thank you for taking a look at the PR and for your helpful feedback.

File: azurerm/resource_arm_policy_definition.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+		return fmt.Errorf("Error deleting Policy Definition %q: %+v", name, err)
+	}
+
+	return nil
+}
+
+func parsePolicyDefinitionNameFromId(id string) (string, error) {
+	idURL, err := url.ParseRequestURI(id)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jaymitre) [reply azurerm/resource_arm_policy_definition.go]: When we use parseAzureResourceID we get an error that a resource group isn't in the id string.

File: website/azurerm.erb [hunk 2]
```diff
@@ -594,6 +594,15 @@
               </ul>
             </li>
 
+            <li<%= sidebar_current("docs-azurerm-policy-definition") %>>
+              <a href="#">Policy Definition</a>
```
Comment (tombuildsstuff) [inline website/azurerm.erb]: missed this earlier - can we make this `Policy Resources`?

File: website/docs/r/policy_definition.html.markdown [hunk 3]
```diff
@@ -0,0 +1,89 @@
+---
+layout: "azurerm"
+page_title: "Azure Resource Manager: azure_policy_definition"
```
Comment (tombuildsstuff) [inline website/docs/r/policy_definition.html.markdown]: can we update `azure_policy_definition` -> `azurerm_policy_definition`

**Comment by jaymitre** [comment]
Updated the naming.  Thank you.

File: azurerm/resource_arm_policy_definition.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_policy_definition.go]: in which case can we return just the last segment of this? given we can guarantee there's only 6 segments in the URI (once it's split based on `/`) we should be able to make this:

```
components := strings.Split(path, "/")

if len(components) != 6 {
  return nil, fmt.Errorf("Azure Policy Definition Id should have 6 segments, got %d: '%s'", len(components), path)
}

return components[5]
```

File: azurerm/resource_arm_policy_definition.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] Error reading Policy Definition %q - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error reading Policy Definition %+v", err)
+	}
+
+	d.Set("id", resp.ID)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_policy_definition.go]: we don't need to set `id` here - since it's already set in the Create method above

**Comment by jaymitre** [comment]
Hey @tombuildsstuff, made a few changes with to address your remaining comments.

File: website/docs/r/policy_definition.html.markdown [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the policy definition. Changing this forces a
+    new resource to be created.
+
+* `policy_type` - (Required) The policy type.  The value can be BuiltIn, Custom 
+    or NotSpecified. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/policy_definition.html.markdown]: _minor_ can we quote these values e.g. "`BuiltIn`, `Custom` and `NotSpecified`"

File: website/docs/r/policy_definition.html.markdown [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+* `mode` - (Required) The policy mode that allows you to specify which resource
+    types will be evaluated.  The value can be All, Indexed or
+    NotSpecified. Changing this resource forces a new resource to be
```
Comment (tombuildsstuff) [inline website/docs/r/policy_definition.html.markdown]: (same here) can we quote these values e.g. `All`, `Indexed` or `NotSpecified`

File: website/docs/r/policy_definition.html.markdown [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    a then block.
+
+* `parameters` - (Optional) Parameters for the policy definition. This field
+    is a json object that allows you to parameterize your policy definition.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The policy defenition id.
```
Comment (tombuildsstuff) [inline website/docs/r/policy_definition.html.markdown]: _minor_ `defenition` -> `definition`

File: website/docs/r/policy_definition.html.markdown [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+## Import
+
+Policy Definitions can be imported using the `policy name`, e.g.
+
+```shell
+terraform import azurerm_policy_definition.testPolict  /subscriptions/<SUBSCRIPTION_ID>/providers/Microsoft.Authorization/policyDefinitions/<POLICY_NAME>
```
Comment (tombuildsstuff) [inline website/docs/r/policy_definition.html.markdown]: _minor_ could we update `testPolict` to be `testPolicy`? :)

**Comment by tombuildsstuff** [comment]
hey @jaymitre 

Thanks for pushing those updates - [I've opened a PR on your fork](https://github.com/jaymitre/terraform-provider-azurerm/pull/1) (since I don't have permissions to push to your fork's master branch) which includes a couple of bug fixes I noticed when testing this PR - if we can merge that in then this PR should be good to merge :)

Thanks!

**Comment by jaymitre** [comment]
@tombuildsstuff Thank you for taking the time to review the PR and make those fixes.  Your change has been merged.

**Comment by tombuildsstuff** [comment]
Tests pass:

```
$ acctests azurerm TestAccAzureRMPolicyDefinition_
=== RUN   TestAccAzureRMPolicyDefinition_importBasic
--- PASS: TestAccAzureRMPolicyDefinition_importBasic (26.40s)
=== RUN   TestAccAzureRMPolicyDefinition_basic
--- PASS: TestAccAzureRMPolicyDefinition_basic (15.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	41.519s
```

**Comment by JasonNguyenTX** [comment]
Hello, does it support for policy tier?  Currently it seems that the default assignment tier is free.  I would like to see if it can support standard tier

**Pull request merged**